### ISSUE TITLE
Dynamic setup

### ISSUE THREAD

**Issue by pociej** [nicolodavis/boardgame.io#695]
Let say i have a game that could have many setups and expected setup is stored outside and should be loaded into game. As example lets say i have card game for 4 players where i want to have control of cards distribution. So there is big set of possible distribution but i want to be able  to write client side. 

```
<Client gameID={someGameId} distributionId={myDistributionId}/> 
```

and then serverside have something like this : 

```
const server = Server({
  games: [Game],
});
```
where Game is not an object by async function that take params passed on lient (`distributionId in this case` ) and return initial setup. Or maybe better is to keep things server side as they are 
but allow setup in `Game` object 

```  setup: (ctx, setupData) => G,```

to be asyc.

**Comment by pociej** [comment]
@nicolodavis @delucis what do you think?

**Comment by delucis** [comment]
I can see the use case, but am a little hesitant about the two approaches you describe as the game definition is designed to be run on both client and server. If `setup` was async, the APIs it would use would probably end up being different in those two environments. For example you might want to use `fetch` in the browser, but that’s not native to Node, or read from disk or a database on the server in a way you wouldn’t do in the browser.

For purely client-side apps, it should be possible to fetch and use data already:

```js
const gameFactory = asyncConfig => ({
  setup: ctx => ({ asyncConfig }),
});

const asyncResource = await fetch('https://resource.url/');
const data = await asyncResource.json();
const game = gameFactory(data);
```

This same request data could theoretically be sent from the client to the Lobby API as `setupData`, but I agree that it would be better to have a way to do that server side. Once 0.40.0 is released, it will be possible to access the API router directly. This will make it easier to customise the API with middleware, so I think something like this would be possible:

```js
const server = Server({ /* options */ });

// add middleware to the create game route to set `setupData` with async data
server.router.use('/games/:name/create', async ctx => {
  // read some options sent from the client
  const opts = ctx.request.body.options;
  // fetch your async data
  const asyncData = await fetchDataFromSomeAPI(opts);
  // set setupData in the request body to be used by the createGame code
  ctx.request.body.setupData = asyncData;
});

server.run(8000);
```

How does that approach sound for your use case?

**Comment by pociej** [comment]
@delucis 

I played with this and i think i found the way. Let me know what do you think, so the patch wiuld be that, on client i would allow one more option : 

```
 Client({
    game: Game
    board: Board,
    numPlayers: 4,
    multiplayer: SocketIO({ server : myUrl })
    setupId,  //on client only id is known, not fullsetup as some informations has to be hidden 
  });
```

then this setupId is passed down to the server in socket io. 
Server side we normall do : 

```
const server = Server({
  games: [myGameFactory],
});
```

duting initialization if not the object by factory is provided then server would call this factory passing `setupId` as param. It doesnt solve asyc yet but at least allow sync which is enough for me ( however i will check, maybe it will work with async too )

**Comment by delucis** [comment]
That’s a very specific pattern that would likely only work for a few use cases — we should support all kinds of arbitrary setup data (not just an ID) — and we already have one mechanism for game customisation via `setupData`, so I’m not sure creating a second one with game factories is a great idea. (I’m also not sure it would work technically, as while the `setup` function can use `setupData`, the server then expects all instances of a game to use the same game code.)

Is there a reason the approaches I described wouldn’t work for you?

**Comment by pociej** [comment]
@delucis

1) Im not running pure cient game so we can forget about first 
2) 0.40 is not released yet 
3) seems that rest api is used only by lobby, i mean when i use client side `Client(myConfig)` it is connected to server via socket.io and the API router is not called at all.

**Comment by delucis** [comment]
Hi Grzegorz, it’ll be a little while before we get 0.40.0 ready as there are some pretty major changes still required. And yes, this relies on the REST API and calling the `create` route. It’s an extra step, but I think it’s the cleanest way for us to support complex match creation flows.

It’s true that we currently support a kind of automatic match creation from the socket server when clients connect with an unknown match ID and the game is initialized without needing the Lobby API:

https://github.com/nicolodavis/boardgame.io/blob/d7284254c106996a538b6bdfdfd2078e6f52a2eb/src/master/master.ts#L381-L384

I’m not personally a huge fan of this implicit match creation and the resulting match won’t have metadata, so is not protected by credentials etc. We could potentially allow clients to pass `setupData` here, but I don’t see a straightforward way to support async fetching. That said, if you do want to propose any improvements in this area, please feel free!

**Comment by nicolodavis** [comment]
I agree that the implicit match creation is a bit ugly. It was mostly in place because we didn't have a Game Creation REST API in the early days. I think there is a TODO [here](https://github.com/nicolodavis/boardgame.io/blob/master/src/master/master.ts#L382) to get rid of it.

Even if we decide to keep it, we should allow passing `setupData` like we do for the REST API.

**Comment by Rupesh-ark** [comment]
Closing this old thread: the approach discussed above is fully supported — `Server()` returns its Koa `router`, so you can add middleware on the create route that fetches async data and injects it as `setupData` before match creation. Async `setup` itself isn't planned since game code must run synchronously on both client and server.
